GoogleUtilities/Tests/Unit/Environment/GULHeartbeatDateStorageTest.m

```diff
@@ -304,6 +304,62 @@ - (void)testConformsToHeartbeatStorableProtocol {
   XCTAssertTrue([self.storage conformsToProtocol:@protocol(GULHeartbeatDateStorable)]);
 }
 
+#pragma mark - Concurrency tests
+
+- (void)testConcurrentReadWriteWithSingleInstance {
+  dispatch_queue_t concurrentQueue = dispatch_queue_create(
+      "testConcurrentReadWriteToTheSameFileFromDifferentInstances", DISPATCH_QUEUE_CONCURRENT);
+
+  NSUInteger attemptsCount = 50;
+
+  XCTestExpectation *expectation = [self expectationWithDescription:self.name];
+  expectation.expectedFulfillmentCount = attemptsCount;
+
+  for (NSUInteger i = 0; i < attemptsCount; i++) {
+    dispatch_async(concurrentQueue, ^{
+      [self assertWriteAndReadNoFileCorruption:self.storage];
+      [expectation fulfill];
+    });
+  }
+
+  [self waitForExpectations:@[ expectation ] timeout:1];
+}
+
+- (void)testConcurrentReadWritesToTheSameFileFromDifferentInstances {
+  dispatch_queue_t concurrentQueue = dispatch_queue_create(
+      "testConcurrentReadWriteToTheSameFileFromDifferentInstances", DISPATCH_QUEUE_CONCURRENT);
+
+  GULHeartbeatDateStorage *storage1 =
+      [[GULHeartbeatDateStorage alloc] initWithFileName:kTestFileName];
+  GULHeartbeatDateStorage *storage2 =
+      [[GULHeartbeatDateStorage alloc] initWithFileName:kTestFileName];
+
+  NSUInteger attemptsCount = 50;
+
+  XCTestExpectation *expectation1 =
+      [self expectationWithDescription:[NSString stringWithFormat:@"%@-1", self.name]];
+  expectation1.expectedFulfillmentCount = attemptsCount;
+
+  for (NSUInteger i = 0; i < attemptsCount; i++) {
+    dispatch_async(concurrentQueue, ^{
+      [self assertWriteAndReadNoFileCorruption:storage1];
+      [expectation1 fulfill];
+    });
+  }
+
+  XCTestExpectation *expectation2 =
+      [self expectationWithDescription:[NSString stringWithFormat:@"%@-2", self.name]];
+  expectation2.expectedFulfillmentCount = attemptsCount;
+  for (NSUInteger i = 0; i < attemptsCount; i++) {
+    dispatch_async(concurrentQueue, ^{
```

Comment:
Should this use a different queue than storage1?

Reply:
Since the queue is created with `DISPATCH_QUEUE_CONCURRENT` parameter, the blocks will be executed concurrently, not sequentially, so creating another queue looks unnecessary to me.